### PR TITLE
Generate dash input as clearlydefined bom

### PR DESCRIPTION
**Pull request by kse3hi**
ADO#14132 - Generate dash input as clearlydefined bom

### CODE REVIEW

File: .devcontainer/scripts/configure-proxies.sh [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+# This file is maintained by velocitas CLI, do not modify manually. Change settings in .velocitas.json
```
Comment (doosuu) [inline .devcontainer/scripts/configure-proxies.sh]: In this repo, this file is not maintained by the CLI, line needs to be removed

**Comment by doosuu** [comment]
> ADO#14132

Better description needed. An ADO ticket ID is not enough.

File: README.md [hunk 2]
```diff
@@ -28,6 +28,10 @@ Shall the action fail if a license violation is detected? Default `false`
 
 **Required** Path to the config file for the license checker.
 
+### `generate-dash`
+
+Generate clearlydefined.input based on [ClearDefined Scheme](https://docs.clearlydefined.io/using-data)
```
Comment (doosuu) [inline README.md]:
```suggestion
Generate clearlydefined.input based on [ClearDefined Schema](https://docs.clearlydefined.io/using-data)
```

The linked website does not mention anything about `clearlydefined.input`.  Also the parameter is called `generate-dash` how does this match the description?

File: src/action.py [hunk 3]
```diff
@@ -157,10 +164,14 @@ def main():
         generate_notice_file(
             origin_to_licenses, f"{github_workspace}/{notice_file_path}"
         )
-        notice_file_is_dirty = is_dirty(github_workspace, notice_file_path)
-        if notice_file_is_dirty:
-            output_update_hint(github_workspace, notice_file_path)
-            workflow_failure = True
+        # notice_file_is_dirty = is_dirty(github_workspace, notice_file_path)
```
Comment (doosuu) [inline src/action.py]: Uncomment this block

File: src/action.py [hunk 4]
```diff
@@ -54,6 +55,12 @@ def get_args():
         help="Pass GitHub token to overcome possible rate limiting issues",
     )
 
+    parser.add_argument(
+        "generate_dash",
+        type=lambda x: bool(strtobool(x)),
+        help="Generate DEPENDENCY based on ClearDefined IDs",
```
Comment (doosuu) [inline src/action.py]:
```suggestion
        help="Generate Eclipse Dash compliant input file",
```

File: src/action.py [hunk 5]
```diff
@@ -157,10 +164,14 @@ def main():
...
-        if notice_file_is_dirty:
-            output_update_hint(github_workspace, notice_file_path)
-            workflow_failure = True
+        # notice_file_is_dirty = is_dirty(github_workspace, notice_file_path)
+        # if notice_file_is_dirty:
+        #     output_update_hint(github_workspace, notice_file_path)
+        #     workflow_failure = True
+
+    if args.generate_dash:
+        print("Generate dash input")
```
Comment (doosuu) [inline src/action.py]:
```suggestion
        print("Generating Eclipse Dash compliant input file")
```

File: src/dash/dashgenerator.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+# *
+# * SPDX-License-Identifier: Apache-2.0
+# ********************************************************************************/
+
+"""Methods to generate a notice file from given dependencies."""
+
+from licensevalidator.lib.dependency import DependencyInfo
+
+def generate_dependency(
+    path_to_clearlydefined_input: str, origin_to_dependencies: dict[str, list[DependencyInfo]]
```
Comment (doosuu) [inline src/dash/dashgenerator.py]:
```suggestion
    path_to_output_file: str, origin_to_dependencies: dict[str, list[DependencyInfo]]
```

File: src/dash/dashgenerator.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+) -> None:
+    """Generate a notice file from the given dependencies in 
+        type/provider/namespace/name/revision format"""
+    with open(path_to_clearlydefined_input, "w", encoding="utf-8") as file:
+        for origin, dep_infos in origin_to_dependencies.items():
+            for dep_info in dep_infos:
+                if origin == "Python":
```
Comment (doosuu) [inline src/dash/dashgenerator.py]: Introduce `get_type_provider_namspace_prefix` method to increase extendability in the future.

```
def get_type_provider_namespace_prefix(origin: str) -> string:
   mapping = {
    "Python": "pypi/pypi/-/",
    "Workflows": "git/github/",
    "Rust": "crate/cratesio/-/"
   }
   return mapping[origin]
```

File: src/dash/dashgenerator.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+                    file.write(f"pypi/pypi/-/{dep_info.name}/{dep_info.version}\n")
+                if origin == "Workflows":
+                    file.write(f"git/github/{dep_info.name}/{dep_info.version}\n")
+                if origin == "Rust":
+                    file.write(f"crate/cratesio/-/{dep_info.name}/{dep_info.version}\n")
+                else:
+                    print(f"Unkwown origin: {origin}")
+    return
```
Comment (doosuu) [inline src/dash/dashgenerator.py]: Omit return statement when returning None

File: src/dash/dashgenerator.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+# * http://www.apache.org/licenses/LICENSE-2.0
+# *
+# * SPDX-License-Identifier: Apache-2.0
+# ********************************************************************************/
+
+"""Methods to generate a notice file from given dependencies."""
+
+from licensevalidator.lib.dependency import DependencyInfo
+
+def generate_dependency(
```
Comment (doosuu) [inline src/dash/dashgenerator.py]: Give this method a better name, i.e. `generate_eclipse_dash_input_file`

File: testbench/multilang/.github/workflows/my-other-workflow.yaml
Comment (doosuu) [inline testbench/multilang/.github/workflows/my-other-workflow.yaml:1]: Why was this file added?

File: testbench/multilang/.github/workflows/my-workflow.yml
Comment (doosuu) [inline testbench/multilang/.github/workflows/my-workflow.yml:1]: Why was this file added?

**Review by doosuu — CHANGES_REQUESTED** [review]
Please see my comments

File: src/dash/dashgenerator.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        for origin, dep_infos in origin_to_dependencies.items():
+            for dep_info in dep_infos:
+                if origin == "Python":
+                    file.write(f"pypi/pypi/-/{dep_info.name}/{dep_info.version}\n")
+                if origin == "Workflows":
+                    file.write(f"git/github/{dep_info.name}/{dep_info.version}\n")
+                if origin == "Rust":
+                    file.write(f"crate/cratesio/-/{dep_info.name}/{dep_info.version}\n")
+                else:
+                    print(f"Unkwown origin: {origin}")
```
Comment (doosuu) [inline src/dash/dashgenerator.py]: C++/Conan not supported?

File: src/dash/dashgenerator.py (hunk 7 above)
Comment (kse3hi) [reply src/dash/dashgenerator.py]: Implemented and added verification for unknow origin

File: testbench/multilang/.github/workflows/my-workflow.yml
Comment (kse3hi) [reply testbench/multilang/.github/workflows/my-workflow.yml:1]: I would like to ensure in one test run that all possible languages/technology are processed:
cpp
rust
python
**workflows**

File: src/dash/dashgenerator.py (hunk 10 above)
Comment (kse3hi) [reply src/dash/dashgenerator.py]: Added Conan parse. Unfortunately cannot find for now proper prefixes. Probably we will be a first, who use it with cpp :)

File: README.md [hunk 11]
```diff
@@ -28,6 +28,10 @@ Shall the action fail if a license violation is detected? Default `false`
 
 **Required** Path to the config file for the license checker.
 
+### `generate-dash`
```
Comment (doosuu) [inline README.md:31]: As a reader its not clear how `dash` and `clearlydefined` work together. Please provide a better explaination how these 2 are connected.

File: src/dash/dashgenerator.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+# * http://www.apache.org/licenses/LICENSE-2.0
+# *
+# * SPDX-License-Identifier: Apache-2.0
+# ********************************************************************************/
+
+"""Methods to generate a notice file from given dependencies."""
+
+from licensevalidator.lib.dependency import DependencyInfo
+
+def generate_clearlydefined_input_file(
```
Comment (doosuu) [inline src/dash/dashgenerator.py]: `Clearlydefined` is the specification for a single line of this output file. The tool that accepts this multiline file of strings in clearlydefined format is Eclipse dash, also the file itself is called `dashgenerator`. I would recommend we also call this method `generate_dash_input_file`.

**Review by doosuu — CHANGES_REQUESTED** [review]
Some minor findings left

File: src/dash/dashgenerator.py (hunk 12 above)
Comment (kse3hi) [reply src/dash/dashgenerator.py]: @doosuu , it is initial implementation of input file, that is ClearlyDefined input. Next stage will be extend it with dash tool usage that will generate DEPENDENCIES file, that will be only dash specific format.

Dash under hood performs verification in front of 2 datasource - 
1) ClearlyDefined one - https://clearlydefined.io/
2) Eclipse Specific one (defined curations) - https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/tree/master

Dash License Tool itself will only care about existence of Issues, that lead to curation creation in case if component "false-positive" due to verification in front of ClearlyDefined datasource. 

Therefore I would like to keep this name for function, to be able distinguish real Dash Tool implementation later

**Review by doosuu — APPROVED** [review]
LGTM

File: action.yml [hunk 13]
```diff
@@ -29,6 +29,10 @@ inputs:
   github-token:
     description: GitHub token to overcome API rate limiting  
     default: ${{ github.token }}
+  generate-dash:
+    description: Generate clearlydefined.input based on ClearDefined Scheme
+    required: false
+    defualt: false
```
Comment (doosuu) [inline action.yml]:
```suggestion
    default: false
```
